core/templates/pages/topic-editor-page/services/topic-editor-state.service.ts

```diff
@@ -18,528 +18,544 @@
+import { UndoRedoService } from 'domain/editor/undo_redo/undo-redo.service.ts';
+import { StorySummary, StorySummaryBackendDict } from 'domain/story/story-summary.model';
 import { TopicRights } from 'domain/topic/topic-rights.model';
+import { SubtopicPageObjectFactory, SubtopicPage, SubtopicPageBackendDict } from 'domain/topic/SubtopicPageObjectFactory.ts';
+import { TopicObjectFactory, Topic, TopicBackendDict } from 'domain/topic/TopicObjectFactory.ts';
+import { TopicRightsBackendApiService } from 'domain/topic/topic-rights-backend-api.service.ts';
+import { RubricObjectFactory, RubricBackendDict, Rubric } from 'domain/skill/RubricObjectFactory.ts';
+import { EditableStoryBackendApiService } from 'domain/story/editable-story-backend-api.service.ts';
+import { EditableTopicBackendApiService } from 'domain/topic/editable-topic-backend-api.service.ts';
+import { AlertsService } from 'services/alerts.service';
+import { SkillSummaryBackendDict } from 'domain/skill/skill-summary.model';
+import { SkillIdToDescriptionMap } from 'domain/topic/SubtopicObjectFactory';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class TopicEditorStateService {
+  constructor(
+    private alertService: AlertsService,
+    private editableStoryBackendApiService: EditableStoryBackendApiService,
+    private editableTopicBackendApiService: EditableTopicBackendApiService,
+    private rubricObjectFactory: RubricObjectFactory,
+    private subtopicPageObjectFactory: SubtopicPageObjectFactory,
+    private topicObjectFactory: TopicObjectFactory,
+    private topicRightsBackendApiService: TopicRightsBackendApiService,
+    private undoRedoService: UndoRedoService) {}
+
+  private _topic: Topic = this.topicObjectFactory.createInterstitialTopic();
+  private _topicRights: TopicRights = TopicRights.createInterstitialRights();
+  // The array that caches all the subtopic pages loaded by the user.
+  private _cachedSubtopicPages: SubtopicPage[] = [];
+  // The array that stores all the ids of the subtopic pages that were not
+  // loaded from the backend i.e those that correspond to newly created
+  // subtopics (and not loaded from the backend).
+  private _newSubtopicPageIds: string[] = [];
+  private _subtopicPage: SubtopicPage =
+  this.subtopicPageObjectFactory.createInterstitialSubtopicPage();
+  private _topicIsInitialized: boolean = false;
+  private _topicIsLoading: boolean = false;
+  private _topicIsBeingSaved: boolean = false;
+  private _topicWithNameExists: boolean = false;
+  private _topicWithUrlFragmentExists: boolean = false;
+  private _canonicalStorySummaries: StorySummary[] = [];
+  private _skillIdToRubricsObject: Record<string, Rubric[]> = {} ;
+  private _skillQuestionCountDict:
+  { [skillId: string]: number; } | undefined[] = {};
+  private _groupedSkillSummaries = {
+    current: [],
+    others: []
+  };
+  private _classroomUrlFragment = 'staging';
+  private _storySummariesInitializedEventEmitter = new EventEmitter();
```

Comment:
Do this for the other three event emitters below.
```suggestion
  private _storySummariesInitializedEventEmitter: EventEmitter<void> = new EventEmitter();
```

Reply:
Done